File: algorithms/Subsets.java
package algorithms;

import java.util.*;

public class Subsets {
    public static List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        result.add(new ArrayList<>());
        for (int num : nums) {
            int size = result.size();
            for (int i = 0; i < size; i++) {
                List<Integer> subset = new ArrayList<>(result.get(i));
                subset.add(num);
                result.add(subset);
            }
        }
        return result;
    }

    public static void main(String[] args) {
        int[] subsetInput = { 1, 2, 3 };
        List<List<Integer>> subsets = subsets(subsetInput);
        System.out.println("All possible subsets for input " + Arrays.toString(subsetInput) + ": " + subsets);
        // output: All possible subsets for input [1, 2, 3]: [[], [1], [2], [1, 2],
        // [3],[1, 3], [2, 3], [1, 2, 3]]
    }

    public static List<List<Integer>> subsets1(int[] nums) {
        return null;
    }
}

File: algorithmsJS/Subsets.js
class Subsets {
  static subsets(nums) {
    const result = [[]];
    for (let num of nums) {
      const size = result.length;
      for (let i = 0; i < size; i++) {
        const subset = result[i].slice();
        subset.push(num);
        result.push(subset);
      }
    }
    return result;
  }

  static subsets1(nums) {
    return result;
  }
}

// Example usage:
const subsetInput = [1, 2, 3];
const subsets = Subsets.subsets(subsetInput);
const subsets1 = Subsets.subsets1(subsetInput);
console.log(
  `All possible subsets for input ${JSON.stringify(subsetInput)}:`,
  subsets
);
console.log(
  `All possible subsets for input ${JSON.stringify(subsetInput)}:`,
  subsets1
);
